sql: TestCancelQueriesRace failed

sql.TestCancelQueriesRace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9108939?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9108939?buildTab=artifacts#/) on master @ [11fbbca584e3bd774ebee3ee94c49ca30a096b26](https://github.com/cockroachdb/cockroach/commits/11fbbca584e3bd774ebee3ee94c49ca30a096b26):


```
=== RUN   TestCancelQueriesRace
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/b287de7194e17a0386ee1775f0d198ba/logTestCancelQueriesRace12901595
    test_log_scope.go:79: use -show-logs to present logs inline
    show_test.go:1213: -- test log scope end --
    show_test.go:1213: Leaked goroutine: goroutine 1437970 [IO wait]:
        internal/poll.runtime_pollWait(0x7f490ae9bf28, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0x89
        internal/poll.(*pollDesc).wait(0xc012f4dc80?, 0xc005696020?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x32
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0xc012f4dc80, {0xc005696020, 0x1, 0x200})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x25a
        net.(*netFD).Read(0xc012f4dc80, {0xc005696020?, 0x7f9765?, 0x435d8d?})
        	GOROOT/src/net/fd_posix.go:55 +0x29
        net.(*conn).Read(0xc00f4b6a20, {0xc005696020?, 0xc005059c18?, 0x436d16?})
        	GOROOT/src/net/net.go:183 +0x45
        io.ReadAtLeast({0x5f1d4c0, 0xc00f4b6a20}, {0xc005696020, 0x1, 0x200}, 0x1)
        	GOROOT/src/io/io.go:332 +0x9a
        io.ReadFull(...)
        	GOROOT/src/io/io.go:351
        github.com/lib/pq.(*conn).ssl(0xc005696000, 0xc0191ba7e0?)
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:1108 +0x16a
        github.com/lib/pq.(*conn).cancel(0xc0072b2f00, {0x5f4ee90, 0xc0191ba7e0})
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:162 +0x2b4
        github.com/lib/pq.(*conn).watchCancel.func1()
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:126 +0x128
        created by github.com/lib/pq.(*conn).watchCancel
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:105 +0xf5
--- FAIL: TestCancelQueriesRace (5.79s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCancelQueriesRace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25545

sql.TestCancelQueriesRace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9477538?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9477538?buildTab=artifacts#/) on master @ [af563f8897aac5c8751566345e83c2c008a887dc](https://github.com/cockroachdb/cockroach/commits/af563f8897aac5c8751566345e83c2c008a887dc):


```
=== RUN   TestCancelQueriesRace
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/8234b428640d8e621ebf8a7f34fa9dcc/logTestCancelQueriesRace2845412301
    test_log_scope.go:79: use -show-logs to present logs inline
    show_test.go:1213: -- test log scope end --
    show_test.go:1213: Leaked goroutine: goroutine 59983 [IO wait]:
        internal/poll.runtime_pollWait(0xffff78b6f938, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0xa0
        internal/poll.(*pollDesc).wait(0x40178e2080?, 0x4017769520?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x28
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0x40178e2080, {0x4017769520, 0x1, 0x200})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x1e8
        net.(*netFD).Read(0x40178e2080, {0x4017769520?, 0x40187f7bd8?, 0x433194?})
        	GOROOT/src/net/fd_posix.go:55 +0x28
        net.(*conn).Read(0x400d43c490, {0x4017769520?, 0x401af89c08?, 0x8e6c1c?})
        	GOROOT/src/net/net.go:183 +0x34
        io.ReadAtLeast({0x58bb820, 0x400d43c490}, {0x4017769520, 0x1, 0x200}, 0x1)
        	GOROOT/src/io/io.go:332 +0xa8
        io.ReadFull(...)
        	GOROOT/src/io/io.go:351
        github.com/lib/pq.(*conn).ssl(0x4017769500, 0x4017769500?)
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:1108 +0x134
        github.com/lib/pq.(*conn).cancel(0x4016fa9800, {0x58ed6c8, 0x40177e1500})
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:162 +0x1f0
        github.com/lib/pq.(*conn).watchCancel.func1()
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:126 +0x10c
        created by github.com/lib/pq.(*conn).watchCancel
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:105 +0xf0
--- FAIL: TestCancelQueriesRace (6.17s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCancelQueriesRace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestCancelQueriesRace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9711411?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9711411?buildTab=artifacts#/) on master @ [0ffba819991a026246029b405abc8fd08ae8d746](https://github.com/cockroachdb/cockroach/commits/0ffba819991a026246029b405abc8fd08ae8d746):


```
=== RUN   TestCancelQueriesRace
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/d07d02991d5ddeff08ff06650d30220a/logTestCancelQueriesRace1956447024
    test_log_scope.go:79: use -show-logs to present logs inline
    show_test.go:1213: -- test log scope end --
    show_test.go:1213: Leaked goroutine: goroutine 133086 [IO wait]:
        internal/poll.runtime_pollWait(0xffff727cab88, 0x72)
        	GOROOT/src/runtime/netpoll.go:305 +0xa0
        internal/poll.(*pollDesc).wait(0x40057b7300?, 0x40114cd820?, 0x0)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x28
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0x40057b7300, {0x40114cd820, 0x1, 0x200})
        	GOROOT/src/internal/poll/fd_unix.go:167 +0x1e8
        net.(*netFD).Read(0x40057b7300, {0x40114cd820?, 0x4001e71bd8?, 0x433194?})
        	GOROOT/src/net/fd_posix.go:55 +0x28
        net.(*conn).Read(0x4003d04518, {0x40114cd820?, 0x402054bc08?, 0x8e6c1c?})
        	GOROOT/src/net/net.go:183 +0x34
        io.ReadAtLeast({0x58ef3c0, 0x4003d04518}, {0x40114cd820, 0x1, 0x200}, 0x1)
        	GOROOT/src/io/io.go:332 +0xa8
        io.ReadFull(...)
        	GOROOT/src/io/io.go:351
        github.com/lib/pq.(*conn).ssl(0x40114cd800, 0x400f283c20?)
        	github.com/lib/pq/external/com_github_lib_pq/conn.go:1108 +0x134
        github.com/lib/pq.(*conn).cancel(0x401c916900, {0x59213a8, 0x400f283c20})
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:162 +0x1f0
        github.com/lib/pq.(*conn).watchCancel.func1()
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:126 +0x10c
        created by github.com/lib/pq.(*conn).watchCancel
        	github.com/lib/pq/external/com_github_lib_pq/conn_go18.go:105 +0xf0
--- FAIL: TestCancelQueriesRace (6.70s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCancelQueriesRace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/show_test.go
+++ b/pkg/sql/show_test.go
@@ -1202,12 +1202,14 @@ func TestCancelQueriesRace(t *testing.T) {
 		_, _ = sqlDB.ExecContext(ctx, `SELECT pg_sleep(10)`)
 		close(waiter)
 	}()
-	_, _ = sqlDB.ExecContext(ctx, `CANCEL QUERIES (
+	_, err := sqlDB.ExecContext(ctx, `CANCEL QUERIES (
 		SELECT query_id FROM [SHOW QUERIES] WHERE query LIKE 'SELECT pg_sleep%'
 	)`)
-	_, _ = sqlDB.ExecContext(ctx, `CANCEL QUERIES (
+	require.NoError(t, err)
+	_, err = sqlDB.ExecContext(ctx, `CANCEL QUERIES (
 		SELECT query_id FROM [SHOW QUERIES] WHERE query LIKE 'SELECT pg_sleep%'
 	)`)
+	require.NoError(t, err)
 
 	cancel()
 	<-waiter

--- a/pkg/sql/tests/BUILD.bazel
+++ b/pkg/sql/tests/BUILD.bazel
@@ -122,7 +122,6 @@ go_test(
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",
         "//pkg/util/tracing",
-        "//pkg/util/tracing/tracingpb",
         "//pkg/util/uuid",
         "@com_github_cockroachdb_cockroach_go_v2//crdb",
         "@com_github_cockroachdb_cockroach_go_v2//crdb/crdbpgx",

--- a/pkg/sql/tests/autocommit_extended_protocol_test.go
+++ b/pkg/sql/tests/autocommit_extended_protocol_test.go
@@ -16,6 +16,7 @@ import (
 	"errors"
 	"net/url"
 	"strings"
+	"sync/atomic"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
@@ -27,7 +28,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/jackc/pgx/v4"
 	"github.com/stretchr/testify/require"
 )
@@ -158,14 +158,14 @@ func TestErrorDuringExtendedProtocolCommit(t *testing.T) {
 	var db *gosql.DB
 
 	var shouldErrorOnAutoCommit syncutil.AtomicBool
-	var traceID tracingpb.TraceID
+	var traceID atomic.Uint64
 	params, _ := CreateTestServerParams()
 	params.Knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
 		DisableAutoCommitDuringExec: true,
 		BeforeExecute: func(ctx context.Context, stmt string, descriptors *descs.Collection) {
 			if strings.Contains(stmt, "SELECT 'cat'") {
 				shouldErrorOnAutoCommit.Set(true)
-				traceID = tracing.SpanFromContext(ctx).TraceID()
+				traceID.Store(uint64(tracing.SpanFromContext(ctx).TraceID()))
 			}
 		},
 		BeforeAutoCommit: func(ctx context.Context, stmt string) error {
@@ -174,7 +174,7 @@ func TestErrorDuringExtendedProtocolCommit(t *testing.T) {
 				// saw when executing our test query. This is so we know that this
 				// autocommit corresponds to our test qyery rather than an internal
 				// query.
-				if traceID == tracing.SpanFromContext(ctx).TraceID() {
+				if traceID.Load() == uint64(tracing.SpanFromContext(ctx).TraceID()) {
 					shouldErrorOnAutoCommit.Set(false)
 					return errors.New("injected error")
 				}

--- a/pkg/util/leaktest/leaktest.go
+++ b/pkg/util/leaktest/leaktest.go
@@ -62,6 +62,8 @@ func interestingGoroutines() map[int64]string {
 			strings.Contains(stack, "github.com/jackc/pgconn.(*PgConn).asyncClose.func1") ||
 			// Ignore pgconn which creates a goroutine to watch context cancellation.
 			strings.Contains(stack, "github.com/jackc/pgconn/internal/ctxwatch.(*ContextWatcher).Watch.func1") ||
+			// Ignore pq goroutine that watches for context cancellation.
+			strings.Contains(stack, "github.com/lib/pq.(*conn).watchCancel") ||
 			// Seems to be gccgo specific.
 			(runtime.Compiler == "gccgo" && strings.Contains(stack, "testing.T.Parallel")) ||
 			// Ignore intentionally long-running logging goroutines that live for the

